src/plot.js

```diff
@@ -96,12 +95,38 @@ export function plot(options = {}) {
       .call(applyInlineStyles, style)
     .node();
 
-  for (const mark of marks) {
+  // A mark layout is responsible for:
+  // * applying the scales, possibly pushing a new one (for legends)
+  // * returning a filtered index and values to render, possibly new channels to rescale
+  const newScaleChannels = new Map();
+  const layouts = [];
+  const newOptions = {};
+  for (let i = 0; i < marks.length; ++i) {
+    const mark = marks[i];
+    const {channels} = layouts[i] = applyLayout.call(mark, mark.layout, [markIndex.get(mark)], scaleDescriptors, markChannels.get(mark), dimensions, options);
+    if (channels) {
+      for (const [, {scale, options: hint}] of channels) {
+        if (scale) newOptions[scale] = {...hint, ...options[scale]};
+      }
+      addScaleChannels(newScaleChannels, channels, newOptions);
+      markChannels.set(mark, channels);
+    }
+  }
+
+  // compute descriptors for new scales
+  const newScaleDescriptors = Scales(newScaleChannels, newOptions);
+  scaleDescriptors = {...newScaleDescriptors, ...scaleDescriptors};
```

Comment:
When we’re constructing new scales here, we have to be careful because it’s possible that a layout can define new channels for a scale that already existed and had existing channels. This code only considers the new channels when creating new scales, but there might be some existing channels, too—we need to combine the new and old channels together for each scale that needs to be reconstructed after computing the layouts.

Reply:
On this question I took the view that applying a layout would not recompute the values for the older channels (they can operate on their mark only), and would not be allowed to modify an existing scale once it's been instantiated (in order to avoid any loop if one layout derives color from x, and the other derives x from color). Maybe overly simplistic?

Reply:
I agree we should not recompute channel values. The current code in this branch _does_ modify an existing scale, though, or at least it overwrites it with a new scale derived from the new channels and ignoring the old channels. I’m proposing instead that it overwrites it with a new scale derived from the union of old and new channels, just like if you added another mark to a plot.

Reply:
Well, it’s not quite working like I expect. I wonder what I’m missing. For the hexbinDot example, there’s an old channel associated with the *r* scale that represents the input to the hexbin:

```js
[3750, 3800, 3250, NaN, 3450, 3650, 3625, 4675, 3475, 4250, …]
```

And the new channel associated with the *r* scale represents the output of the hexbin:

```js
[24, 25, 26, 4, 7, 4, 3, 10, 1, 1, 23, …]
```

The old channel probably shouldn’t be associated with the *r* scale since its unscaled. (The numbers represent the mass of penguins in grams, not radii in pixels.) I need to dig in a bit deeper to figure out what’s going on.

Reply:
Okay, I think the way to think about this is that the *r* channel associated with the dot mark is getting *replaced* by the hexbin layout. So we don’t want to union the two channels here—we just want the new channel. But we’ll need to carefully track which channels are getting replaced (same name, same mark) to determine the surviving set of channels associated with each scale.